pennylane/labs/vibrational/bosonic_mapping.py

```diff
@@ -0,0 +1,160 @@
+
+
+@singledispatch
+def _binary_mapping_dispatch(bose_operator, d, ps=False, wires_map=None, tol=None):
+    """Dispatches to appropriate function if bose_operator is a BoseWord or BoseSentence."""
+    raise ValueError(f"bose_operator must be a BoseWord or BoseSentence, got: {bose_operator}")
+
+
+@_binary_mapping_dispatch.register
+def _(bose_operator: BoseWord, d, ps=False, wire_map=None, tol=None):
+
+    if d < 2:
+        raise ValueError(f"Number of bosonic states cannot be less than 2, provided {d}.")
+    nqub_per_boson = int(np.ceil(np.log2(d)))
+
+    cr_mat = np.zeros((d, d))
+    for s in range(d - 1):
+        cr_mat[s + 1, s] = np.sqrt(s + 1.0)
+
+    d_mat = {"+": cr_mat, "-": cr_mat.T}
+
+    qubit_operator = PauliSentence({PauliWord({}): 1.0})
+
+    for item in bose_operator.items():
+        (_, b_idx), sign = item
+        op = PauliSentence()
+        sparse_dmat = np.nonzero(d_mat[sign])
+        for i, j in zip(*sparse_dmat):
+            coeff = d_mat[sign][i][j]
+
+            binary_row = list(map(int, bin(i)[2:]))[::-1]
+            if nqub_per_boson > len(binary_row):
+                binary_row += [0] * (nqub_per_boson - len(binary_row))
+
+            binary_col = list(map(int, bin(j)[2:]))[::-1]
+            if nqub_per_boson > len(binary_col):
+                binary_col += [0] * (nqub_per_boson - len(binary_col))
+
+            pauliOp = PauliSentence({PauliWord({}): 1.0})
+            for n in range(nqub_per_boson):
+                pauliOp @= _get_pauli_op(binary_row[n], binary_col[n], n + b_idx * nqub_per_boson)
+
+            op += coeff * pauliOp
+        qubit_operator @= op
+    qubit_operator.simplify()
+
+    for pw in qubit_operator:
+        if tol is not None and abs(qml.math.imag(qubit_operator[pw])) <= tol:
+            qubit_operator[pw] = qml.math.real(qubit_operator[pw])
+
+    wires = list(bose_operator.wires) or [0]
+    identity_wire = wires[0]
+    if not ps:
+        qubit_operator = qubit_operator.operation(wire_order=[identity_wire])
+
+    if wire_map:
+        return qubit_operator.map_wires(wire_map)
+
+    return qubit_operator
+
+
+@_binary_mapping_dispatch.register
+def _(bose_operator: BoseSentence, d, ps=False, wire_map=None, tol=None):
+
+    qubit_operator = PauliSentence()
+
+    for bw, coeff in bose_operator.items():
+        bose_word_as_ps = binary_mapping(bw, d=d, ps=True)
+
+        for pw in bose_word_as_ps:
+            qubit_operator[pw] = qubit_operator[pw] + bose_word_as_ps[pw] * coeff
+
+            if tol is not None and abs(qml.math.imag(qubit_operator[pw])) <= tol:
+                qubit_operator[pw] = qml.math.real(qubit_operator[pw])
+
+    qubit_operator.simplify(tol=1e-16)
+
+    wires = list(bose_operator.wires) or [0]
+    identity_wire = wires[0]
+    if not ps:
+        qubit_operator = qubit_operator.operation(wire_order=[identity_wire])
+
+    if wire_map:
+        return qubit_operator.map_wires(wire_map)
```

Comment:
This code is duplicated between the two functions. Can we move this into binary_mapping instead to reduce code duplication?